RTL support for TheAdmin theme

hishamco opened the pull request:
Fixes #3015

hishamco commented:
@agriffard looking for elegant way to run `rtlcss` instead of `gulp rtl`

agriffard commented:
I tried the PR:
Launch and setup a Blog site.
Enable the Localization module.
Add the `ar` culture.
Select it as the default culture.

=> In the admin layout, the `@Orchard.CultureName()` stays at `fr` which is my System culture.

Do you remember when you forked OrchardCore on your repo?

hishamco commented:
It's a bit old, but I remember that I rebase to one of the earlier commits .. let me check again

hishamco commented:
As I can see the `rtl` branch is 3 commits ahead of `OrchardCMS:dev`.

hishamco commented:
> => In the admin layout, the @Orchard.CultureName() stays at fr which is my System culture.

@agriffard could you try to restart the application, so the added culture will take the affect

agriffard commented:
Even if I restart the application, it is not working.

I can only see it if I suffix the url by something like this: `?culture=ar-YE`

hishamco commented:
Strange!! just review the RTL layout and please answer my question in the first thread, then we will check why the localization isn't working properly

agriffard commented:
After adding the arabic .po files, it works as expected:
![image](https://user-images.githubusercontent.com/703248/55981663-82747c80-5c97-11e9-8343-3360775e6796.png)

Login Page:
![image](https://user-images.githubusercontent.com/703248/55981702-97511000-5c97-11e9-9850-e1b9a14895e3.png)

I don't know how you can use something else than `gulp rtlcss` to build the -rtl css file.

hishamco commented:
Strange!! it worked at my end without adding Arabic.po file this is just for localise the resources nothing else .. it's great to see RTL shine in Orchard Core, but we need to investigate on the localization issue that you come up with

hishamco commented:
@khoshroomahdi could you please try out Persian (Farsi) at your end and check how the RTL layout is look like, we need more than one to convince the others that Orchard Core is RTL ready 😁

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Admin/Views/AdminDashboard.cshtml:
```diff
@@ -1,4 +1,4 @@
 <div class="alert alert-success">
     <h4>@T["Welcome to Orchard"]</h4>
-    <p>@T["Feel free to browse the menu on the left and discover all its possibilities."]</p>
+    <p>@T["Feel free to browse the menu and discover all its possibilities."]</p>
```
Maybe it should just be a different translation for rtl languages ;)

hishamco replied:
May be, but I remove it to avoid the confusing

sebastienros commented on gulpfile.js:
```diff
@@ -30,6 +31,14 @@ require("events").EventEmitter.prototype._maxListeners = 100;
 */
 
+gulp.task("rtl", function() {
+    var styleFolder = "./src/OrchardCore.Themes/TheAdmin/wwwroot/Styles";
```
No. We can't have a specific reference to this css in this file.
Also why only this file and not all the other ones?

hishamco replied:
That's why I asked above, I'm not gulp expert, so we need a better way for doing this

sebastienros commented:
Why not check what the RTLCSS did, and apply all the changes on the `rtl` selector that is defined in `<body>` only once, and then when we do a change in TheAdmin we need to ensure it also works for RTL?

hishamco commented:
> Why not check what the RTLCSS did, and apply all the changes on the rtl selector that is defined in <body> only once, and then when we do a change in TheAdmin we need to ensure it also works for RTL?

I already did that few days ago, it's not easy as you adding `rtl` selector to the body, it brings a new style sheet that you need to inject

What I did is the simple way to support the RTL, except we need an elegant way to run the gulp task

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Admin/Views/AdminDashboard.cshtml:
```diff
@@ -1,4 +1,5 @@
 <div class="alert alert-success">
     <h4>@T["Welcome to Orchard"]</h4>
-    <p>@T["Feel free to browse the menu and discover all its possibilities."]</p>
+    <p>@T[$"Feel free to browse the menu on the {(OrchardHelper.CultureDir() == "ltr" ? "left" : "right")} and discover all its possibilities."]</p>
```
It's worse than anything. A translated string can't be dynamic, we can't find the values automatically. And even if in this case there are only two possibilities, it's bad. Either remove the term as you did earlier, or leave it as it was initially and we can handle it in a localization entry.

hishamco replied:
So that's why I remove it at the beginning :)

hishamco commented:
@sebastienros as we discusses yesterday on Slack, I spent few hours to get this done with a single style file with no luck :( but I'm so close .. the problem is i'm not a gulp expert :smilely

I think your suggestion is very good for now.
- I added `generateRTL` to `Assets.json` to avoid hard coded names in gulp file
- I modified the `LinkTagHelper`, so support RTL will be super easy

I will push my changes ASAP

hishamco commented:
The only thing that I need to find somehow is why the `rtl` is not generate the RTL files when both TheAdmin.css and TheAdmin.min.css are exist

hishamco commented:
It works fine when I run `gulp rebuild` if that is the right behavior the PR is ready from my side

hishamco commented:
Finally I did it 😄 .. I spent several hours to make both language direction into a single style, believe it or not most of the time wasted on `css-prefix` which is limited and another package that I am using and finally discovered there was an error in the docs 😣
What I did is the following:
- Remove the earlier changes in the `LinkTagHelper` it is not needed anymore
- Modify the gulp file so the RTL will run during the css pipeline generation
- Modify the `CultureDir()` so it should return ltr instead of empty string

Hopefully @sebastienros will be supper happy because he ~~insists~~ likes that both language direction styles should go into a single file 😀

hishamco commented:
Seems I made a change accidently on the login page when I was working on PR #3518 it moved to here while login is related to admin

sebastienros commented on Package.json:
```diff
@@ -17,16 +17,19 @@
     "typescript": "3.2.2"
   },
   "dependencies": {
+    "gulp-rtlcss": "^1.3.0",
```
I don't think you need this dependency anymore.

sebastienros commented on src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/LinkTagHelper.cs:
```diff
@@ -1,4 +1,5 @@
-﻿using System;
+using System;
+using System.Globalization;
```
Change not necessary

sebastienros commented:
There are also merge conflict, so you need to rebase your branch (or merge dev), ignore any change in TehAdmin (for simplicity of merging) then regenerate the css files (gulp rebuild) and commit again. Then I'll merge

hishamco commented:
Sure

hishamco commented:
Seems you did most what I did locally .. but I'm facing Internet issues .. thanks

hishamco commented:
I think it's time to merge

sebastienros commented:
it is, but I am doing some tests, currently facing issues with crowdin configuration, on it

aghili371 commented:
it has a problem in media popup and assets, the still left to right and the tree will hide.

![2](https://user-images.githubusercontent.com/24512878/57099579-0fea4000-6d1d-11e9-8fb9-ba179eb5cce4.JPG)
![Uploading 1.JPG…]()

hishamco commented:
@aghili371 could you please file an issue so it can easily tracked, I will try to fix it as soon as possible.

I'm not sure if it's inline or internal style because it will not work with the current implementation
